File: backend/app/services/provenance.py
"""Source provenance & availability (Connectivity Intelligence, #5).

Two distinct concepts (spec §3, §7, §8):

- **Provenance** — *where* a persisted piece of evidence actually came from. Stamped
  onto ``Source.meta["provenance"]`` at collection time and never inferred after the
  fact. A source is ``LIVE_WEB`` **only** when an external provider returned it during
  the current run; a cache reuse is ``CACHED_WEB``; local retrieval is ``LOCAL_*``.

- **Availability** — the *display state* derived from provenance + freshness, shown as
  a badge so a reader can see at a glance whether evidence is live, cached, local or
  stale. This is derived, not stored, so it always reflects current freshness.

``UNAVAILABLE`` is deliberately **not** a persisted-source state: a persisted ``Source``
is evidence we actually hold. A source that could not be fetched has no row; it is
counted at the run-health level instead (so we never invent phantom "unavailable"
sources — spec §34).
"""
from __future__ import annotations

# --- Provenance: where the evidence came from ------------------------------ #
LIVE_WEB = "live_web"
CACHED_WEB = "cached_web"
LOCAL_DOCUMENT = "local_document"
LOCAL_MEMORY = "local_memory"
LOCAL_DATABASE = "local_database"

# --- Availability: display state (spec §3) --------------------------------- #
LIVE = "live"
CACHED = "cached"
LOCAL = "local"
STALE = "stale"
UNAVAILABLE = "unavailable"
UNKNOWN = "unknown"

# External source types are fetched over the network; the rest are local.
EXTERNAL_SOURCE_TYPES = frozenset({"web", "docs", "news", "community", "github", "papers"})
LOCAL_SOURCE_TYPES = frozenset({"documents"})

# Default provenance for a source type when nothing more specific was stamped.
_DEFAULT_PROVENANCE = {
    "documents": LOCAL_DOCUMENT,
}


def is_external(source_type: str) -> bool:
    """True for source types that require a network fetch (web/docs/news/community/
    github/papers). Local types (documents) are False."""
    return source_type in EXTERNAL_SOURCE_TYPES


def default_provenance(source_type: str) -> str:
    """The provenance to assume for a source type absent an explicit stamp. Local
    types are local; external types are treated as live (they only exist as rows
    because a live fetch produced them)."""
    return _DEFAULT_PROVENANCE.get(source_type, LIVE_WEB if is_external(source_type) else LOCAL_DATABASE)


def provenance_of(source_type: str, meta: dict | None) -> str:
    """Read the stamped provenance from a source's meta, falling back to the type
    default. Never guesses "live" for something that wasn't actually fetched live —
    the default for external types reflects that an external row *is* a live result
    unless explicitly stamped ``cached_web``."""
    prov = (meta or {}).get("provenance")
    if prov in (LIVE_WEB, CACHED_WEB, LOCAL_DOCUMENT, LOCAL_MEMORY, LOCAL_DATABASE):
        return prov
    return default_provenance(source_type)


def availability_of(provenance: str, freshness: str) -> str:
    """Derive the display availability from provenance + freshness (spec §3).

    Freshness wins when evidence exists but is stale; otherwise provenance maps
    straight through. This keeps "cached recently" distinct from "live" and
    "uploaded today" distinct from "published today".
    """
    if freshness == "stale":
        return STALE
    if provenance == CACHED_WEB:
        return CACHED
    if provenance in (LOCAL_DOCUMENT, LOCAL_MEMORY, LOCAL_DATABASE):
        return LOCAL
    if provenance == LIVE_WEB:
        return LIVE
    return UNKNOWN


# Research-health labels for a whole run (spec §22).
FULLY_LIVE = "fully_live"
PARTIALLY_DEGRADED = "partially_degraded"
CACHE_ASSISTED = "cache_assisted"
LOCAL_ONLY = "local_only"
EXTERNAL_UNAVAILABLE = "external_unavailable"


def research_health(
    *, live: int, cached: int, local: int, unavailable: int
) -> str:
    """Summarise a run's overall source health into one label (spec §22).

    Honest by construction: ``fully_live`` only when nothing was cached, local-only or
    unavailable; ``external_unavailable`` only when external evidence was requested but
    none (live or cached) came back.
    """
    if unavailable and not live and not cached and not local:
        return EXTERNAL_UNAVAILABLE
    if live and not cached and not unavailable and not local:
        return FULLY_LIVE
    if not live and not cached and local:
        return LOCAL_ONLY
    if cached and not live:
        return CACHE_ASSISTED
    if unavailable or cached:
        return PARTIALLY_DEGRADED
    if live:
        return FULLY_LIVE
    return LOCAL_ONLY

File: frontend/src/lib/provenance.ts
// Pure presentation helpers for source provenance/availability (#5). Kept free of
// React so the mapping (provenance/availability -> label/badge) is unit-testable,
// mirroring lib/evidence.ts.
import type { Availability, Provenance, ResearchHealth } from "../api/types";

export const AVAILABILITY_META: Record<
  Availability,
  { label: string; dot: string; className: string }
> = {
  live: { label: "Live", dot: "🟢", className: "bg-emerald-100 text-emerald-700" },
  cached: { label: "Cached", dot: "🔵", className: "bg-sky-100 text-sky-700" },
  local: { label: "Local", dot: "📄", className: "bg-violet-100 text-violet-700" },
  stale: { label: "Stale", dot: "🟡", className: "bg-amber-100 text-amber-700" },
  unavailable: { label: "Unavailable", dot: "🔴", className: "bg-rose-100 text-rose-700" },
  unknown: { label: "Unknown", dot: "⚪", className: "bg-slate-100 text-slate-500" },
};

export function availabilityMeta(a: string) {
  return AVAILABILITY_META[a as Availability] ?? AVAILABILITY_META.unknown;
}

const PROVENANCE_TO_AVAILABILITY: Record<Provenance, Availability> = {
  live_web: "live",
  cached_web: "cached",
  local_document: "local",
  local_memory: "local",
  local_database: "local",
};

/**
 * Derive the display availability from provenance + freshness. Mirrors the backend
 * `availability_of` — freshness "stale" wins, otherwise provenance maps through. Used
 * as a fallback when the backend `availability` field isn't present.
 */
export function deriveAvailability(provenance: string, freshness: string): Availability {
  if (freshness === "stale") return "stale";
  return PROVENANCE_TO_AVAILABILITY[provenance as Provenance] ?? "unknown";
}

export const RESEARCH_HEALTH_META: Record<
  ResearchHealth,
  { label: string; dot: string; className: string }
> = {
  fully_live: { label: "Fully live", dot: "🟢", className: "bg-emerald-100 text-emerald-700" },
  partially_degraded: {
    label: "Partially degraded",
    dot: "🟡",
    className: "bg-amber-100 text-amber-700",
  },
  cache_assisted: { label: "Cache-assisted", dot: "🔵", className: "bg-sky-100 text-sky-700" },
  local_only: { label: "Local only", dot: "📄", className: "bg-violet-100 text-violet-700" },
  external_unavailable: {
    label: "External sources unavailable",
    dot: "🔴",
    className: "bg-rose-100 text-rose-700",
  },
};

export function researchHealthMeta(h: string) {
  return (
    RESEARCH_HEALTH_META[h as ResearchHealth] ?? {
      label: h || "Unknown",
      dot: "⚪",
      className: "bg-slate-100 text-slate-500",
    }
  );
}

export const CONNECTIVITY_META: Record<
  string,
  { label: string; dot: string; className: string }
> = {
  online: { label: "Online", dot: "🟢", className: "border-emerald-200 bg-emerald-50 text-emerald-700" },
  recovering: { label: "Recovering", dot: "🟢", className: "border-emerald-200 bg-emerald-50 text-emerald-700" },
  degraded: { label: "Degraded", dot: "🟡", className: "border-amber-200 bg-amber-50 text-amber-700" },
  local_only: { label: "Local only", dot: "📄", className: "border-violet-200 bg-violet-50 text-violet-700" },
  offline: { label: "Offline", dot: "🔴", className: "border-rose-200 bg-rose-50 text-rose-700" },
  unknown: { label: "Unknown", dot: "⚪", className: "border-slate-200 bg-slate-50 text-slate-500" },
};

export function connectivityMeta(state: string) {
  return CONNECTIVITY_META[state] ?? CONNECTIVITY_META.unknown;
}
